lib/glueby/contract/token.rb

```diff
@@ -0,0 +1,191 @@
+      # @raise [InvalidTokenType] if token is not reissuable.
+      def reissue!(issuer:, amount:)
+        raise Glueby::Contract::Errors::InvalidAmount unless amount.positive?
+        raise Glueby::Contract::Errors::InvalidTokenType unless token_type == Tapyrus::Color::TokenTypes::REISSUABLE
+
+        estimated_fee = FixedFeeProvider.new.fee(Tapyrus::Tx.new)
+        funding_script = Tapyrus::Script.parse_from_payload(@payload)
+        funding_tx = create_funding_tx(wallet: issuer, amount: estimated_fee, script: funding_script)
+        tx = create_reissue_tx(funding_tx: funding_tx, issuer: issuer, amount: amount, color_id: color_id)
+        [funding_tx, tx].each { |tx| Glueby::Internal::RPC.client.sendrawtransaction(tx.to_payload.bth) }
+      end
+
+      # Send the token to other wallet
+      #
+      # @param sender [Glueby::Wallet] wallet to send this token
+      # @param receiver [Glueby::Wallet] wallet to receive this token
+      # @param amount [Integer]
+      # @return [Token] receiver token
+      # @raise [InsufficientFunds] if wallet does not have enough TPC to send transaction.
+      # @raise [InsufficientTokens] if wallet does not have enough token to send.
+      # @raise [InvalidAmount] if amount is not positive integer.
+      def transfer!(sender:, receiver:, amount: 1)
+        raise Glueby::Contract::Errors::InvalidAmount unless amount.positive?
+
+        tx = create_transfer_tx(color_id: color_id, sender: sender, receiver: receiver, amount: amount)
+        Glueby::Internal::RPC.client.sendrawtransaction(tx.to_payload.bth)
+      end
+
+      # Burn token
+      # If amount is not specified or 0, burn all token associated with the wallet.
+      #
+      # @param sender [Glueby::Wallet] wallet to send this token
+      # @param amount [Integer]
+      # @raise [InsufficientFunds] if wallet does not have enough TPC to send transaction.
+      # @raise [InsufficientTokens] if wallet does not have enough token to send transaction.
+      # @raise [InvalidAmount] if amount is not positive integer.
+      def burn!(sender:, amount: 0)
+        raise Glueby::Contract::Errors::InvalidAmount unless amount.positive?
+
+        tx = create_burn_tx(color_id: color_id, sender: sender, amount: amount)
+        Glueby::Internal::RPC.client.sendrawtransaction(tx.to_payload.bth)
+      end
+
+      # Return balance of token in the specified wallet.
+      # @param wallet [Glueby::Wallet]
+      # @return [Integer] amount of utxo value associated with this token.
+      def amount(wallet:)
+        # collect utxo associated with this address
+        utxos = wallet.internal_wallet.list_unspent
+        _, results = collect_colored_outputs(utxos, color_id)
+        results.sum { |result| result[:amount] }
+      end
+
+      # Return color id
+      # @return [Tapyrus::Color::ColorIdentifier]
+      def color_id
+        Tapyrus::Color::ColorIdentifier.parse_from_payload([@token_type, Tapyrus.sha256(@payload)].pack('Ca*'))
```

Comment:
If you receive a token from another user, maybe there is no payload and only a hashed value?

Reply:
fix at d0977e9

Reply:
... and rename attribute name "payload " to script_pubkey because we need this attribute only when token is reissuable.